designs/http_client_design.md

```diff
@@ -0,0 +1,250 @@
+}
+
+// Production-ready client with FAIR built in.
+type FairHTTPClient struct {
+    client *http.Client
+}
+
+// Configures FairHTTPClient.
+type FairHTTPClientConfig struct {
+    FairConfig      *FairRoundTripperConfig // required
+    Timeout         time.Duration           // default 30s
+    TransportConfig *TransportConfig        // connection pooling defaults
+}
+```
+
+Additional helpers:
+
+* `type ClientIDExtractor func(*http.Request) ([]byte, error)`
+* `HeaderClientIDExtractor(headerName string) ClientIDExtractor`
+* `ContextClientIDExtractor(key any) ClientIDExtractor`
+* `HashedClientIDExtractor(inner ClientIDExtractor) ClientIDExtractor`
+
+### Behavior
+
+**Request flow**
+
+On every request:
+
+1. **Extract client ID** with `ClientIDExtractor` (header by default, e.g. `X-Client-ID`).
+
+   * If extraction fails and `FallbackClientID` is set, use fallback.
+   * If extraction fails and no fallback, return an error immediately.
+
+2. **Register with FAIR**
+
+   * Call `Tracker.RegisterRequest(ctx, clientID)`.
+
+3. **Throttle if needed**
+
+   * If `ShouldThrottle` is true:
+     * Option A (default): return a synthetic HTTP response with configurable status (default 429), `X-Fair-Throttled: true`, optional `Retry-After`.
+     * Call `OnThrottle` hook if set.
+   * No network call is made in this case.
+
+4. **Execute request** using the wrapped `http.RoundTripper`.
+
+5. **Report outcome**
+   * Map `(resp, err)` into `request.Outcome` (success vs resource failure) and call `Tracker.ReportOutcome(ctx, clientID, outcome)`.
+
+**Retry interaction**
+
+* Retries are **optional**, configured via `RetryConfig`.
+* Default behavior:
+  * Retry only **idempotent methods** (`GET`, `HEAD`, `OPTIONS`, `TRACE`, `PUT`, `DELETE`).
+  * Retry on:
+    * Network errors.
+    * Configurable 5xx codes (default: 500, 502, 503, 504) and optionally 429.
+  * Use exponential backoff with jitter, with small, conservative defaults (e.g. max 3 attempts).
+
+> Important: **Each retry attempt** registers and reports its own outcome with FAIR, so client behavior remains visible to the fairness algorithm and can be throttled if necessary.
+
+**Standalone client**
+
+`FairHTTPClient`:
+
+* Constructs an `http.Transport` with production defaults:
+  * Higher `MaxIdleConns` / `MaxIdleConnsPerHost` than `stdlib` defaults.
+  * Sensible connection and header timeouts.
+* Wraps that transport with `FairRoundTripper`.
+* Exposes:
+  * `Do(req *http.Request) (*http.Response, error)`
+  * Convenience methods `Get`, `Post`, etc., delegating to the underlying `http.Client`.
+
+This lets users choose:
+
+* **Wrapper usage** (minimal change to existing code), or
+* **New client usage** (get FAIR + better transport defaults in one go).
+
+**Fail-Open vs Fail-Closed Behavior**
+
+If the client is unable to communicate with the FAIR tracker (e.g., local process issue, shared-memory corruption, or unexpected internal errors), we must choose between:
```

Comment:
Nit - Also possible that  `FAIR` servers are themselves unavailable (crashes/overloads) (or) perceived as unavailable network (network)